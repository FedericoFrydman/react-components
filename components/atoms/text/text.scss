@import '../../assets/styles/fonts.scss';
@import '../../assets/styles/colors.scss';

.lsux-text {
  margin: 0;

  // Text Weights
  &--normal {
    font-weight: normal;
  }
  &--bold {
    font-weight: bold;
  }
  &--semibold {
    font-weight: 600;
  }

  // Text Sizes
  &--tiny {
    font-size: 10px;
    font-style: normal;
    letter-spacing: -0.01em;
    line-height: 12px;
  }
  &--description,
  &--small {
    font-size: 12px;
    font-style: normal;
    letter-spacing: -0.006em;
    line-height: 16px;
  }
  &--body,
  &--medium {
    font-size: 14px;
    font-style: normal;
    letter-spacing: -0.006em;
    line-height: 18px;
  }
  &--large {
    font-size: 16px;
    font-style: normal;
    letter-spacing: -0.011em;
    line-height: 22px;
  }
  &--extra-large {
    font-size: 18px;
    font-style: normal;
    letter-spacing: -0.014em;
    line-height: 24px;
  }

  // Text Styles
  &--italic {
    font-style: italic;
  }
  &--strike {
    text-decoration: line-through;
  }

  &--no-margin {
    margin: 0;
  }

  &--separator {
    align-items: center;
    display: flex;
    text-transform: lowercase;

    &::before,
    &::after {
      background-color: $N300;
      content: '';
      flex-grow: 1;
      height: 1px;
    }

    &::before {
      margin-right: 16px;
    }

    &::after {
      margin-left: 16px;
    }
  }
}
