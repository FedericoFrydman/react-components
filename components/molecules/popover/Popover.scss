@import '../../assets/styles/colors.scss';

.lsux-popover {
  background: $N00;
  border: 1px solid $N300;
  border-radius: var(--brand-border-radius);
  position: absolute;
  width: 288px;
  z-index: 100;

  &:after,
  &:before {
    border: solid transparent;
    content: '';
    height: 0;
    pointer-events: none;
    position: absolute;
    width: 0;
  }

  &:after {
    border-color: rgba(0, 0, 0, 0);
  }

  &:before {
    border-color: rgba(0, 0, 0, 0);
  }

  &.top:after,
  &.top:before {
    bottom: 100%;
    left: 50%;
  }

  &.top:after {
    border-bottom-color: $N00;
    border-width: 10px;
    margin-left: calc(-1 * 10px);
  }

  &.top:before {
    border-bottom-color: $N300;
    border-width: calc(10px + 1px + 1px);
    margin-left: calc(-1 * (10px + 1px + 1px));
  }

  &.top.top0:after,
  &.top.top0:before {
    left: calc(2 * 10px);
  }

  &.top.top25:after,
  &.top.top25:before {
    left: 25%;
  }

  &.top.top50:after,
  &.top.top50:before {
    left: 50%;
  }

  &.top.top75:after,
  &.top.top75:before {
    left: 75%;
  }

  &.top.top100:after,
  &.top.top100:before {
    left: calc(100% - 2 * 10px);
  }

  &.right:after,
  &.right:before {
    left: 100%;
  }

  &.right:after {
    border-left-color: $N00;
    border-width: 10px;
    margin-top: calc(-1 * 10px);
  }

  &.right:before {
    border-left-color: $N300;
    border-width: calc(10px + 1px + 1px);
    margin-top: calc(-1 * (10px + 1px + 1px));
  }

  &.right.right0:after,
  &.right.right0:before {
    top: calc(2 * 10px);
  }

  &.right.right25:after,
  &.right.right25:before {
    top: 25%;
  }

  &.right.right50:after,
  &.right.right50:before {
    top: 50%;
  }

  &.right.right75:after,
  &.right.right75:before {
    top: 75%;
  }

  &.right.right100:after,
  &.right.right100:before {
    top: calc(100% - 2 * 10px);
  }

  &.left:after,
  &.left:before {
    right: 100%;
  }

  &.left:after {
    border-right-color: $N00;
    border-width: 10px;
    margin-top: calc(-1 * 10px);
  }

  &.left:before {
    border-right-color: $N300;
    border-width: calc(10px + 1px + 1px);
    margin-top: calc(-1 * (10px + 1px + 1px));
  }

  &.left.left0:after,
  &.left.left0:before {
    top: calc(2 * 10px);
  }

  &.left.left25:after,
  &.left.left25:before {
    top: 25%;
  }

  &.left.left50:after,
  &.left.left50:before {
    top: 50%;
  }

  &.left.left75:after,
  &.left.left75:before {
    top: 75%;
  }

  &.left.left100:after,
  &.left.left100:before {
    top: calc(100% - 2 * 10px);
  }

  &.bottom:after,
  &.bottom:before {
    top: 100%;
  }

  &.bottom:after {
    border-top-color: $N00;
    border-width: 10px;
    margin-left: calc(-1 * 10px);
  }

  &.bottom:before {
    border-top-color: $N300;
    border-width: calc(10px + 1px + 1px);
    margin-left: calc(-1 * (10px + 1px + 1px));
  }

  &.bottom.bottom0:after,
  &.bottom.bottom0:before {
    left: calc(2 * 10px);
  }

  &.bottom.bottom25:after,
  &.bottom.bottom25:before {
    left: 25%;
  }

  &.bottom.bottom50:after,
  &.bottom.bottom50:before {
    left: 50%;
  }

  &.bottom.bottom75:after,
  &.bottom.bottom75:before {
    left: 75%;
  }

  &.bottom.bottom100:after,
  &.bottom.bottom100:before {
    left: calc(100% - 2 * 10px);
  }

  &.lsux-popover--dark {
    background: $N800;
    border: 1px solid $N800;

    p {
      color: $N00;
    }

    &.top:after,
    &.top:before {
      border-bottom-color: $N800;
    }

    &.right:after,
    &.right:before {
      border-left-color: $N800;
    }

    &.left:after,
    &.left:before {
      border-right-color: $N800;
    }

    &.bottom:after,
    &.bottom:before {
      border-top-color: $N800;
    }
  }

  .lsux-popover--header {
    display: flex;
    justify-content: space-between;

    .lsux-popover--title {
      font-weight: 500;
    }
  }
}
