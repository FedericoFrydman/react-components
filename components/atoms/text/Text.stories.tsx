import { Story, Meta } from '@storybook/react';

import { Text, TextProps } from './Text';

const TextStory: Story<TextProps> = (args: TextProps) => <Text {...args} />;

export const Default = TextStory.bind({});
Default.args = {
  text: 'This is example body text',
};
export const Tiny = TextStory.bind({});
Tiny.args = {
  text: 'This is example body text',
  textSize: 'tiny',
};
export const Small = TextStory.bind({});
Small.args = {
  text: 'This is example body text',
  textSize: 'small',
};
export const Medium = TextStory.bind({});
Medium.args = {
  text: 'This is example body text',
  textSize: 'medium',
};
export const Large = TextStory.bind({});
Large.args = {
  text: 'This is example body text',
  textSize: 'large',
};
export const ExtraLarge = TextStory.bind({});
ExtraLarge.args = {
  text: 'This is example body text',
  textSize: 'extra-large',
};
export const Bold = TextStory.bind({});
Bold.args = {
  bold: true,
  text: 'This is example body text',
};
export const Semibold = TextStory.bind({});
Semibold.args = {
  semiBold: true,
  text: 'This is example body text',
};
export const Italic = TextStory.bind({});
Italic.args = {
  italic: true,
  text: 'This is example body text',
};
export const Strike = TextStory.bind({});
Strike.args = {
  strike: true,
  text: 'This is example body text',
};
export const WithSeparator = TextStory.bind({});
WithSeparator.args = {
  separator: true,
  text: 'This is example body text',
};
export const Paragraph = TextStory.bind({});
Paragraph.args = {
  as: 'p',
  children: 'A <p> can also be rendered',
};
export const WithChildren = TextStory.bind({});
WithChildren.args = {
  children: 'Some text inside the tags instead of using the text prop',
};

export default {
  argTypes: {
    as: { control: { type: 'select' }, options: ['span', 'p'] },
    bold: { control: { type: 'boolean' } },
    italic: { control: { type: 'boolean' } },
    semiBold: { control: { type: 'boolean' } },
    separator: { control: { type: 'boolean' } },
    strike: { control: { type: 'boolean' } },
    textColor: {
      control: { type: 'select' },
      options: ['N1000', 'N800', 'N700', 'N500', 'N300', 'N100', 'N00'],
    },
    textHeight: {
      control: { type: 'text' },
    },
    textSize: {
      control: { type: 'select' },
      options: ['body', 'description', 'tiny', 'small', 'medium', 'large', 'extra-large'],
    },
    textWeight: { control: { type: 'select' }, options: ['normal', 'semibold', 'bold'] },
  },
  component: TextStory,
  title: 'Typography/Text (Body)',
} as Meta;
