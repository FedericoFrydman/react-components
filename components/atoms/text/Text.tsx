import { forwardRef, ReactElement, Ref } from 'react';
import cn from 'classnames';

import './text.scss';
import '../../assets/styles/utilities.scss';

export type TextAs = 'p' | 'span';

export type TextSize = 'body' | 'description' | 'tiny' | 'small' | 'medium' | 'large' | 'extra-large';

export type TextWeight = 'normal' | 'semibold' | 'bold';

export type TextColor = 'N1000' | 'N800' | 'N700' | 'N500' | 'N300' | 'N100' | 'N00';

export interface TextProps extends Omit<React.HTMLProps<HTMLParagraphElement>, 'css'> {
  /**
   * Render a span or a paragraph. Default 'span'.
   */
  as?: TextAs;
  /**
   * Text value to be rendered
   */
  text?: string;
  /**
   * Defines the text height
   */
  textHeight?: string;
  /**
   * Size of the text. Defaults to `medium`.
   */
  textSize?: TextSize;
  /**
   * Weight of the text
   */
  textWeight?: TextWeight;
  /**
   * Sets the `textWeight` to bold
   * @deprecated Use `fontWeight='bold'` instead. Will be removed in future verison.
   */
  bold?: boolean;
  /**
   * Sets the `textWeight` to semibold
   * @deprecated Use `fontWeight='semibold'` instead. Will be removed in future verison.
   */
  semiBold?: boolean;
  /**
   * Sets the text style to italic.
   */
  italic?: boolean;
  /**
   * Add a strikethrough to the text.
   */
  strike?: boolean;
  /**
   * Text value to be rendered
   */
  children?: string | number | ReactElement;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Color of the text
   * @deprecated Text defaults to N800 color. This prop will be removed in a future verison.
   */
  color?: TextColor;
  /**
   * Modifies the text as separator
   */
  separator?: boolean;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLParagraphElement>;
}

const TextComponent = (
  {
    as,
    bold,
    children,
    color,
    italic,
    semiBold,
    separator,
    strike,
    text,
    textHeight,
    textSize = 'medium',
    textWeight = 'normal',
    classNames = [],
    ...props
  }: TextProps,
  ref: TextProps['ref'],
) => {
  // Warn if bold or semibold props are used
  if (bold || semiBold) {
    console.warn('bold & semibold are deprecated. Please use `textWeight` instead.');
  }

  // Warn if color prop is used
  if (color) {
    console.warn('color is deprecated. Text is defaulted to $N800');
  }

  // Warn if body or description textSize values are used
  if (textSize === 'body' || textSize === 'description') {
    console.warn(
      'textSizes `body` and `description` have been deprecated. Please use `tiny`, `small`, `medium`, `large`, or `extra-large` instead.',
    );
  }

  // Set textWeight
  if (semiBold) {
    textWeight = 'semibold';
  }
  if (bold) {
    textWeight = 'bold';
  }

  const classText = 'lsux-text';
  const setHeight = { height: textHeight, overflow: 'auto' };
  const useHeight = textHeight ? setHeight : {};

  const cnText = cn(
    {
      [`${classText}`]: true,
      [`${classText}--separator`]: separator,
      [`${classText}--italic`]: italic,
      [`${classText}--strike`]: strike,
      [`${classText}--${textSize}`]: textSize,
      [`${classText}--${textWeight}`]: textWeight,
    },
    classNames,
  );

  if (as === 'span') {
    return (
      <span {...props} ref={ref} style={{ ...useHeight, ...props.style }} className={cnText}>
        {text ? text : children}
      </span>
    );
  } else {
    return (
      <p {...props} ref={ref} style={{ ...useHeight, ...props.style }} className={cnText}>
        {text ? text : children}
      </p>
    );
  }
};

export const Text = forwardRef<HTMLParagraphElement, TextProps>(TextComponent);
