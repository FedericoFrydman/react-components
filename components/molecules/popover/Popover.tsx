import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import './Popover.scss';
import '../../assets/styles/shadows.scss';
import { Text } from '../../atoms/text/Text';
import '../../assets/styles/utilities.scss';
import { Icon } from '../../atoms/icon/Icon';

export type PopoverArrowSide = 'top' | 'bottom' | 'left' | 'right';

export type PopoverArrowDistance = '0' | '25' | '50' | '75' | '100';

export type PopoverShadowSide = 'top' | 'bottom' | 'left' | 'right';

export interface PopoverProps extends Omit<React.HTMLProps<HTMLDivElement>, 'css'> {
  /**
   * The children of the arrow box
   */
  children: React.ReactNode;
  /**
   * The side of the arrow box where the arrow exists
   */
  arrowSide: PopoverArrowSide;
  /**
   * The side of the arrow box where the shadow exists
   */
  shadowSide: PopoverShadowSide;
  /**
   * ArrowDistance. The distance from the edge in %.
   */
  arrowDistance?: PopoverArrowDistance;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Dark mode
   */
  dark?: boolean;
  /**
   * Popover title
   */
  title?: string;
  /**
   * Define if popover has a close action
   */
  closeAction?: boolean;
  /**
   * Popover action
   */
  onClick?: (e: React.MouseEvent) => void | unknown;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const PopoverComponent = (
  {
    children,
    arrowSide = 'top',
    shadowSide = 'bottom',
    arrowDistance = '50',
    classNames = [],
    dark,
    title,
    closeAction,
    onClick,
    ...props
  }: PopoverProps,
  ref: PopoverProps['ref'],
) => {
  const iconColor = dark ? 'N00' : 'default';
  const hasHeader = title || closeAction;

  const classPopover = 'lsux-popover';
  const cnPopover = cn(
    {
      ['p-4']: true,
      [`${classPopover}`]: true,
      [`${arrowSide}`]: arrowSide,
      [arrowSide + arrowDistance]: true,
      [`shadow400-${shadowSide}`]: shadowSide,
      [`${classPopover}--dark`]: dark,
    },
    classNames,
  );

  return (
    <div {...props} className={cnPopover} ref={ref}>
      {hasHeader && (
        <div className="lsux-popover--header">
          {title && <Text text={title} as="p" classNames={['lsux-popover--title', 'pb-3']} />}
          {closeAction && <Icon name="action_close" color={iconColor} onClick={onClick} />}
        </div>
      )}
      {children}
    </div>
  );
};

/**
 * An Popover is used to display things like a help balloon or simple popup content
 */

export const Popover = forwardRef<HTMLDivElement, PopoverProps>(PopoverComponent);
