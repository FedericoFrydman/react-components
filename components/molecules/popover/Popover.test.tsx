import { shallow, ShallowWrapper } from 'enzyme';
import { Popover } from './Popover';
import { Text } from '../../atoms/text/Text';

describe('Popover Component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Popover arrowSide="top" shadowSide="bottom">
        <Text text="I am a text child" as="p" />
      </Popover>,
    );
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
