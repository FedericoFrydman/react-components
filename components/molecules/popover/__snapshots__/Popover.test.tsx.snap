// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Component should render the component 1`] = `
<div
  className="p-4 lsux-popover top top50 shadow400-bottom"
>
  <ForwardRef(TextComponent)
    as="p"
    text="I am a text child"
  />
</div>
`;
